packages/core/src/common/message-rpc/rpc-message-encoder.ts

```diff
@@ -84,6 +85,108 @@ export class ResponseError extends Error {
     }
 }
 
+/**
+ * Custom error thrown by the {@link RpcMessageEncoder} if an error occurred during the encoding and the
+ * object could not be written to the given {@link WriteBuffer}
+ */
+export class EncodingError extends Error {
+    constructor(msg: string, public cause?: Error) {
+        super(msg);
+    }
+}
+
+/**
+ * A `RpcMessageDecoder` parses a a binary message received via {@link ReadBuffer} into a {@link RpcMessage}
+ */
+export interface RpcMessageDecoder {
+    parse(buffer: ReadBuffer): RpcMessage;
+}
+
+/**
+ * A `RpcMessageEncoder` writes {@link RpcMessage} objects to a {@link WriteBuffer}. Note that it is
+ * up to clients to commit the message. This allows for multiple messages being
+ * encoded before sending.
+ */
+export interface RpcMessageEncoder {
+    cancel(buf: WriteBuffer, requestId: number): void;
+
+    notification(buf: WriteBuffer, requestId: number, method: string, args: any[]): void
+
+    request(buf: WriteBuffer, requestId: number, method: string, args: any[]): void
+
+    replyOK(buf: WriteBuffer, requestId: number, res: any): void
+
+    replyErr(buf: WriteBuffer, requestId: number, err: any): void
+
+}
+
+export const defaultMsgPack = new MsgPack({ moreTypes: true, encodeUndefinedAsNil: false, bundleStrings: true });
+// Add custom msgpackR extension for ResponseErrors.
+addExtension({
```

Comment:
MsgpackR supports global extensions that enable custom encoding of certain class instances. This approach can also be used in the plugin API to handle the custom encoding of VSCode Uris etc. (which is currently done via JSON replacer/reviver)

Extensions are added globally to msgpack (so they can't be configured per packR instance). For now, i simply add the extension registration here. However, we should probably come up with a better solution like an injectable configuration class.